src/hotspot/share/classfile/classLoaderExt.cpp

```diff
@@ -213,6 +213,12 @@ void ClassLoaderExt::process_jar_manifest(JavaThread* current, ClassPathEntry* e
     char sep = os::file_separator()[0];
     const char* dir_name = entry->name();
     const char* dir_tail = strrchr(dir_name, sep);
+#ifdef _WINDOWS
+    // On Windows, we also support forward slash as the file separator for Class-Path: attribute.
+    if (dir_tail == nullptr) {
+      dir_tail = strrchr(dir_name, '/');
+    }
+#endif
```

Comment:
The value of the Class-Path attribute is a sequence of relative URLs, not file paths, so I wouldn't expect to see any usage of file::file_separator() here, even on Unix systems. Some further work may be needed here to align with how the Class-Path attribute is treated by the application class loader.

Reply:
Given these are required to be relative URLs [1] and given the path component of a URL only uses '/' as a separator [2], then it would seem the use of `os::file_separator` is wrong and it should just be hardcoded to '/' - and further it means the current code has never been correct on Windows. ??

[1] https://docs.oracle.com/en/java/javase/21/docs/specs/jar/jar.html#class-path-attribute
[2] https://en.wikipedia.org/wiki/URL

Reply:
Consider `java -cp dir1/A.jar` ... where the manifiest of A.jar contains:
`Class-Path: a.jar`
The `dir_name` in the above corresponds to `dir1/A.jar`. The `strrchr` is trying to locate the separator between the directory and jar name. Later in the same function, it will combine the directory name (in this case `dir1`) with the names from each of the entry from the attribute and create a class path entry. In the example, it will create a class path entry of `dir1/a.jar`.
Before the patch, on Windows, it only works with '\\' (the default file separator). e.g.` java -cp dir1\A.jar` ... .

Reply:
Also, the modified function is only called during CDS dump time for setting up the shared class paths for the app class loader.

Reply:
To me it sounds like the logic should be always using `/` to locate and then use file_seperator to rewrite to a classpath filesystem entry.

Reply:
No strike that, the logic is more complex.

> Before the patch, on Windows, it only works with '\' (the default file separator). e.g. java -cp dir1\A.jar ... .

So you want to be able to process the classpath if it uses `/` instead of `\` on Windows? Is that how classpath processing normally works on Windows?

The JBS issue says it is fixing the processing of the Class-path attribute, not the processing of the classpath entries. Or maybe I have just misconstrued exactly what this is fixing - the Class-path attribute itself is immaterial here??

Reply:
To better describe what we're fixing, how about changing the JBS title to the following?
`CDS: allow the use of / separator on Windows to locate entries from the Class-Path attribute`

Reply:
How about this?

> ClassLoaderExt::process_jar_manifest() should allow / separator on Windows

Reply:
My question still remains: does regular classpath handling allow both forward- and back-slashes on Windows?

Reply:
Based on Calvin's description, and the definition of the `Class-path` attribute then the code to extract foo.jar should always be using `/` not `os::file_separator`, and then the classpath element should be constructed using `os::file_separator`.

Reply:
The Windows file API allows both `/` and `\`. E.g., both `os::open("foo/bar.jar", ...)` and `os::open("foo\\bar.jar", ...)` work. So in most cases, we just pass the path specified by the user to the OS, without doing any `/` to `\` conversion.
 
The bug in this particular case is: CDS needs to construct the path to a JAR file specified in the ClassPath attribute. As describe in the bug report, the test case is:

```
java -cp foo/bar.jar
```

and bar.jar contains

```
Class-Path: lib.jar
```

CDS needs to construct the string "foo\lib.jar". This is done by

- finding the directory that contains "foo/bar.jar" (which should be "foo")
- appending "\lib.jar" to this directory (which gives us "foo\lib.jar")

The bug is that CDS fails to find the "foo" directory. This is BEFORE any processing of the ClassPath attribute of bar.jar itself.

The fix is -- to be able to find the  directory name, we need to find the longest prefix of the input that ends with either "/" or "\".

Reply:
`foo/bar.jar` is a relative file path. `lib.jar` is a relative URL. On the surface it might appear that they can be combined but there are many reasons why this can't work in general. So if the code is in the VM then it's always going to be approximation to how this JAR file attribute is handled by the application class loader.

Reply:
> > ClassLoaderExt::process_jar_manifest() should allow / separator on Windows

I've updated the titles of the PR and JBS.

Reply:
Just to repeat what I said previously, the value of Class-Path attribute is a sequence of relative URLs. The code in ClassLoaderExt::process_jar_manifest will work for some cases but in general, the appending used in this function can't be worked to combine a file path and URL path.

Reply:
Alan, could you show us some cases where ClassLoaderExt::process_jar_manifest() will fail. IIUC, the failures won't be Windows-specific (i.e., orthogonal to the \ vs / issues in this bug) but rather will happen on all platforms.

Reply:
Can you try examples like `file:lib.jar`, `file:/d/lib.jar`, `/d/lib.jar` for starters. Once you bring Windows file paths into the picture then it's a world of hurt. There is big other topic around escaping where Class-Path is completely broken and we haven't figure out how to fix these issues in a compatibility way.

Reply:
Sorry, I didn't notice last few comments until after pushing the fix. I think if there's any other shortcoming of the function in question, it should be addressed via different bug/RFE.
After a casual search for Class-Path attribute, I found: https://docs.oracle.com/javase/8/docs/technotes/guides/jar/jar.html#classpath
It says the following:
`Currently, the URLs must be relative to the code base of the JAR file for security reasons. Thus, remote optional packages will originate from the same code base as the application.`

Reply:
>  It says the following: `Currently, the URLs must be relative to the code base of the JAR file for security reasons. Thus, remote optional packages will originate from the same code base as the application.`

That's right but there are a lot of issues here that date back to JDK 1.2 when the feature was added. We also have a long standing issue that tooling and plugins in the eco system are putting all manner of junk into the value, which has constrained us from fixing some of the fundamental issues.